datafusion/physical-expr/src/expressions/case.rs

```diff
@@ -148,6 +148,11 @@ impl CaseExpr {
             // Make sure we only consider rows that have not been matched yet
             let when_match = and(&when_match, &remainder)?;
 
+            // When no rows available for when clause, skip then clause
+            if when_match.true_count() == 0 {
+                continue;
+            }
+
```

Comment:
Because even though `when_match` is all false, the `evaluate_selection` still evaluates the expression one time.
https://github.com/apache/arrow-datafusion/blob/827668ab3d4659329ff30e6672f398c934b803be/datafusion/physical-expr/src/expressions/case.rs#L148-L153
so the query like below, will get divide by zero error.
```
select case 1 when 2 the 4/0 end;
```
------
The root cause is that even though `BinaryExpr` receives a `RecordBatch` with 0 rows, it also evaluates once. And, it only appears in `scalar op scalar`.
https://github.com/apache/arrow-datafusion/blob/827668ab3d4659329ff30e6672f398c934b803be/datafusion/physical-expr/src/expressions/binary.rs#L259-L268

Reply:
> so the query like below, will get divide by zero error.
>
> `select case 1 when 2 the 4/0 end;`

Is that case covered by existing tests? If not, can you please add a test?

Reply:
```rust
            // When no rows available for when clause, skip then clause
            if when_match.true_count() == 0 {
                continue;
            }
```
I souble checked that this method accounts for nulls 👍 

https://docs.rs/arrow/latest/arrow/array/struct.BooleanArray.html#method.true_count

Reply:
> Is that case covered by existing tests? If not, can you please add a test?

yes，add this one